### PR TITLE
Responsive, white label-friendly stepper component

### PR DESCRIPTION
**Pull request by tinahollyws**
I've created a responsive stepper component `<ws-stepper />`. It will be used to visualize a timeline-like display of items, and the active item within that range of items.

As with all ws-directives, the screenshots below include the example box which is not part of the component. 

It is white label-friendly, using new utility classes added to Fabric for colours, so implementing it will not introduce white label regressions.

### Large desktop 

image

image

### Medium desktop

image

### Small desktop

image

### Tablet 

image

### Phone

image

### Edge

image

### IE11

There appears to be a rendering issue with IE11 and how it handles flexbox. IE11 considers elements that are positioned absolute—i.e. pulled out of the regular layout flow—in the child elements it arranges in flexbox layouts. This behaviour is not the case in Edge, Firefox, Safari, or Chrome. After speaking with @scompto this is a reasonable and acceptable appearance—it doesn't make it hard to see what item is highlighted due to the active step indicator being offset, and the active item is bolded. Microsoft has fixed this rendering bug in Edge.

image

### CODE REVIEW

File: src/stepper/_ws-stepper.scss [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  background: $color-gray-20;
+  height: 100%;
+  width: $step-line-size;
+  position: absolute;
+  left: $step-indicator-size / 2;
+  z-index: -1;
+
+  @include medium {
+    display: block;
+    height:$step-line-size;
```
Comment (ericjim) [inline src/stepper/_ws-stepper.scss]: add a space after the `:`

File: src/stepper/_ws-stepper.scss [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    margin-bottom: 0;
+  }
+
+  @include medium {
+    margin: 0;
+  }
+}
+
+.ws-stepper__step-indicator--active {
+
```
Comment (ericjim) [inline src/stepper/_ws-stepper.scss]: remove extra line?

**Comment by ericjim** [comment]
Did you try this in an internet explorer browser :D?

File: src/stepper/_ws-stepper.scss [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+.ws-stepper {
+  display: flex;
+
+  @include medium {
+    display: block;
+  }
+}
+
+.ws-stepper__step-indicator-wrapper {
```
Comment (cchawn) [inline src/stepper/_ws-stepper.scss]: We can take advantage of SCSS nesting and stick all of these selectors under the `.ws-stepper` selector like this:

```SCSS
.ws-stepper {
  &__step-indicator-wrapper { ... }
}
```

File: src/stepper/_ws-stepper.scss [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  &:last-of-type {
+    margin-bottom: 0;
+  }
+
+  @include medium {
+    margin: 0;
+  }
+}
+
+.ws-stepper__step-indicator--active {
```
Comment (cchawn) [inline src/stepper/_ws-stepper.scss]: same for nesting these modifiers:
```SCSS
.ws-stepper {
  &__step-indicator {
    ....
    &--active { ... }
  }
}
```

File: src/stepper/_ws-stepper.scss [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  display: flex;
+
+  @include medium {
+    display: block;
+  }
+}
+
+.ws-stepper__step-indicator-wrapper {
+  position: relative;
+  display: flex;
```
Comment (AlexisDeschamps) [inline src/stepper/_ws-stepper.scss]: Maybe order the SCSS statements to be alphabetical like I think we started doing in some other repos

File: src/stepper/ws-stepper.component.test.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    beforeEach(angular.mock.inject(($rootScope, $compile) => {
+      outerScope = $rootScope.$new();
+      element = angular.element(elementHtml);
+      compiledElement = $compile(element)(outerScope);
+      outerScope.$digest();
+    }));
+  }
+
+  describe('steps display', () => {
+    injectElement('<ws-stepper steps="[{ id: 1, title: \'First\' }, { id: 2, title: \'Second\' }, { id: 3, title: \'Third\' }]" active-step="2"/>');
```
Comment (AlexisDeschamps) [inline src/stepper/ws-stepper.component.test.js]: Is is not possible to use \` so `'` can be used within the string without being escaped?

**Comment by DanGodfrey** [comment]
Looks good!

File: src/stepper/ws-stepper.component.test.js (hunk 6 above)
Comment (cchawn) [reply src/stepper/ws-stepper.component.test.js]: I think that our linter doesn't like that, since using the backtick (`) implies a template string

Comment (SethDavenport) [reply src/stepper/ws-stepper.component.test.js]: The linter only likes ` if it's a template string or a multiline string.

I suppose you could do something like:

```js
injectElement(`
  <ws-stepper
    steps="[
     { id: 1, title: 'First' },
     { id: 2, title: 'Second' },
     { id: 3, title: 'Third' }
    ]"
    active-step="2"/>
`);
```

But it's up to you.

Comment (SethDavenport) [reply src/stepper/ws-stepper.component.test.js]: The linter only likes ` if it's a template string or a multiline string.

I suppose you could do something like:

```js
injectElement(`
  <ws-stepper
    steps="[
     { id: 1, title: 'First' },
     { id: 2, title: 'Second' },
     { id: 3, title: 'Third' }
    ]"
    active-step="2"/>
`);
```

But it's up to you.

Comment (tinahollyws) [reply src/stepper/ws-stepper.component.test.js]: Actually, I had the back ticks and I have ESLint run on save and it changed it to that—as @cchawn and @SethDavenport said, the back ticks are for template strings only.

File: src/stepper/_ws-stepper.scss (hunk 3 above)
Comment (tinahollyws) [reply src/stepper/_ws-stepper.scss]: That's true, I could do that, but I find that that looks confusing. Normally when nesting in Sass and using the ampersand it'd be an and-selector (if that makes sense) but this is completely different and actually just uses the parent as prefix as opposed to a shared selector. I would prefer to avoid confusion.

I generally don't like nesting because then we come into CSS specificity issues and have to result to things like `!important`. Even though the nesting you suggested only affects the classname as per above and does not nest the selectors to make it more specific, nesting in general does that, and the readability could be confusing.

**Comment by tinahollyws** [comment]
@ericjim I did not test in Internet Explorer. I probably should!

File: src/stepper/_ws-stepper.scss (hunk 5 above)
Comment (abejfehr) [reply src/stepper/_ws-stepper.scss]: We should add linting for that here to be consistent

File: src/stepper/_ws-stepper.scss (hunk 3 above)
Comment (cchawn) [reply src/stepper/_ws-stepper.scss]: That makes sense! I agree that making things too specific is a headache, but I also think that the power of CSS comes from being able to apply styles based on specificity. So I think that nesting rules is an acceptable practice as long as it doesn't get out of hand (eg: limit the number of levels to 3 deep).

We also tend to use the ampersand solely for combining selectors with the parent, but I don't think we should limit ourselves to that. The ampersand is just a selector for the "immediate parent", so I think it's okay to use it the way I suggested. It would mean "the parent selector *with* some extra stuff on the end". Versus just "the parent selector *and* another selector".

The ampersand is much more powerful than what we use it for. You can write things like:
```SCSS
.foo {
  .bar & { }
  &.baz {}
}
```
which is equivalent to
```CSS
.bar .foo {}
.foo.baz {}
```

But saying all that, I'm not incredibly opinionated since both ways of writing SCSS are readable (imo). I've seen nested BEM selectors in our code base before, but up to you! 😊 

@SethDavenport or @6 do either of you have opinions on this?

Comment (SethDavenport) [reply src/stepper/_ws-stepper.scss]: I nest a lot. But... there are constraints.  This is a broader topic we as a team need to have. Maybe I'll hijack an L&L slot and present something quick and then turn it into a CSS style discussion.

But briefly we have two competing issues with CSS:

* globality
* themability

Globality is that problem where you add a rule to a class and a whole bunch of other stuff gets changed.

Themability is that feature where you want to use custom colour, layout, and typography with our components.

## Globality

So for globality, **nesting is a good thing, as long as it's done right**:  Consider:

`_my-component.scss`

```scss
.ws-my-component {
  // Straight up nested
  h1 {
    /* I've used this a lot but I'm starting to turn away from it. Styles here will not affect our parents, but they will mess up our children unexpectedly. This only half solves the */
  }

  &-title { // equivalent to .ws-my-component-title
    /* Styles here use namespacing - they will most likely not screw up our parent components our our child components. This is a really good thing. But we have to get into a frustrating level of screwiness to override them in a theme situation. */
  }
}

h1 {
 /* really bad. This will mess with any h1 tag in the entire app. This partial should be about my component! */ 
}
```

So my strong preference is for the BEM style namespacing, and nesting syntax makes it (IMO) much easier to read. Honestly I'd do CSS-in-JS if I thought I could get away with it :trollface: because I spent too much time on React.

## Themeability

However, that said, yes: encapsulating styles in components like this makes theming very hard. We end up having to write horrible selectors that drill down the component tree in order to theme stuff.

However this is IMO a feature, not a bug.  Using CSS's globality to do theming is a hack. Even if we do it with special classes, we still end up cluttering our templates.

We are taking steps in the right direction here for themeability with semantic colour names: `primary`, `tertiary`, `fg-`, `bg-` etc. (whether at the class level or the SCSS variable level matters less to me and has more to do with template clutter).

If I had my druthers, the rule of thumb here would be:

1. always use BEM style encapsulation (&-syntax optional but I find it reads better)
2. NEVER use a colour code or a non-semantic colour (e.g. `$color-green-80`). Always use semantic colours only (`$color-primary-dark`, `$colour-success`, etc.).
3. Argue with designers who give you bespoke rainbows and ask them to re-cast that colour as a semantic one. Even if it means coming up with a new semantic.
4. Going forward, expand this philosophy to typescales and space scales.

Unfortunately we're in the process of pushing for the tools we need to do (3) correctly. But there's no reason we can't try and sort out the palette and adopt it going forward.

Comment (SethDavenport) [reply src/stepper/_ws-stepper.scss]: So my thoughts turned out not to be brief at all . 😂  Also completely open to push back.  This is still a bit disorganized in my brain right now but this is the level of conversation we need to have about this I think.

File: src/stepper/_ws-stepper.scss (hunk 5 above)
Comment (tinahollyws) [reply src/stepper/_ws-stepper.scss]: I've made this change, though I think this is something the linter should handle, but it appears as though we aren't linting any styles on this project.

Comment (cchawn) [reply src/stepper/_ws-stepper.scss]: @abejfehr I can add linting in my free time tonight

File: src/stepper/_ws-stepper.scss (hunk 3 above)
Comment (SethDavenport) [reply src/stepper/_ws-stepper.scss]: As for this PR, what you have is fine for now IMO :)

Comment (tinahollyws) [reply src/stepper/_ws-stepper.scss]: I'm personally not a fan of the component-based CSS. Some huge issues can be seen with that in how we componentized the wizard. Wizard styles overwrite Fabric styles. Fabric components in the wizard behave unexpectedly, and solutions like [this](https://github.com/wealthsimple/terminal/blob/W4A-188--line-w-dots/src/app/common/css/modules/_ws-wizard.scss#L153-L169) need to be done to get around unexpected styles in a wizard and restore Fabric styles inside of a component, so I would disagree and say that `.ws-component h1 {}` is **not** a good approach. We are doing this [everywhere in the wizard](https://github.com/wealthsimple/terminal/blob/W4A-188--line-w-dots/src/app/common/css/modules/_ws-wizard.scss#L17) which means you can't just put a `.card` or an `h2` or `p` tag in the wizard and expect it to look the way it normally would throughout the app. So no, I'd say this approach is not good if we want to leverage Fabric.

The `&-title` approach is better and more inline with BEM, which I would agree is a good reasonable standard to use for naming, hence why I'm using it. It's documented, and isn't something we made-up, so it's likely that developers have worked with it already, and if not, we can point them to all of the docs on the internet.

I would disagree that `&__element {}` reads easier but I can get over it. If we want to enforce a style with the linter, I'm cool with following it. 

I also agree with the theming stuff you bring up. I would agree with the SMACSS approach to CSS where you separate your CSS based on concerns, e.g. layout vs. [theme](https://smacss.com/book/type-theme). This makes white labeling efforts easier, though there's a lot of SMACCS that could be overhead that I would avoid. 

We should stop adding colours as you say, and use the classes. Unfortunately because of how we are theming, we should avoid the Sass variables and instead use the [utility classes](https://github.com/wealthsimple/fabric/releases/tag/v3.4.0).

Arguing with designers re: "bespoke rainbows" as you say would definitely avoid problems like [this](https://github.com/wealthsimple/fabric/releases/tag/v3.4.1)—the excess becomes evident when you try to make theme colours into utility classes.

TLDR:

- component-based selectors like `.ws-component h1 {}` are bad
- don't write colours in CSS—use the [utility classes](https://github.com/wealthsimple/fabric/releases/tag/v3.4.0) instead
- BEM naming is good

cc: @SethDavenport @cchawn

Comment (tinahollyws) [reply src/stepper/_ws-stepper.scss]: I'd also be interested in seeing Fabric develop some typescale sort of thing, but that's probably a discussion to have in the Fabric issues rather than ws-directives.

Comment (tinahollyws) [reply src/stepper/_ws-stepper.scss]: What I've done [here](https://github.com/wealthsimple/ws-directives/pull/185/files#diff-ef7693e59ee0dfa86911d2658c02be33R8) I think addresses any themeability issues—instead of specifying colours in the CSS I am using the utility classes I created to toggle the standard colour and the active colour.

Comment (SethDavenport) [reply src/stepper/_ws-stepper.scss]: Yeah I'm with you with `.my-component { h1 { } }`. It provides the illusion of encapsulation but messes with it's children. I'm going to stop doing it.

From this point of view I'm doubling down even further on component specific styling, except for the themeable colours which form a kind of 'api' of sorts. That way we're being explicit about what we consider to be overrideable/themeable; otherwise components hide their layout from each other (even descendants).

All this to say I agree with your TL;DR above.

Comment (SethDavenport) [reply src/stepper/_ws-stepper.scss]: However I'm not 100% sure about the utility classes per se. They work fine, I just don't enjoy having things like `<div class="section mt-2 background-colour-primary ...">` all over my templates. This is a matter of taste so I won't object to it, but I would rather consume the 'themed CSS API' via variables and mixins instead:

`<div class="my-component">`

```scss
.my-component {
  @include background-colour-primary;
  // etc.
}
```

I will say the classes are super handy for dynamic (ng-class) stuff, per your example though.

Comment (tinahollyws) [reply src/stepper/_ws-stepper.scss]: Yes, I agree with this, but not for Terminal for how we're white labeling. We aren't pre-compiling with white label colours available on our machines, which is why I would say don't use it, but [I've included it in Fabric](https://github.com/wealthsimple/fabric/pull/42/files#diff-d8ee409a461718bfb6240710c8c73382R76) so that they can be overridden and used that way by others who choose to use Fabric.

Usage:

```scss
$color-brand-primary: deeppink;

@import 'path/to/fabric';
```

Then any components that use the primary brand colour would have it be whatever override was specified, or the default.

For Terminal, we can't do that though. I agree it mucks up the HTML, but compiling all our white labels in Terminal just won't scale for W4A goals.
